base_local_planner/src/latched_stop_rotate_controller.cpp

```diff
@@ -31,6 +31,13 @@ LatchedStopRotateController::LatchedStopRotateController(const std::string& name
 LatchedStopRotateController::~LatchedStopRotateController() {}
 
 
+bool LatchedStopRotateController::isGoalBypassed(LocalPlannerUtil* planner_util, const geometry_msgs::PoseStamped& global_pose) {
+  std::vector<geometry_msgs::PoseStamped> global_plan;
+  planner_util->getLocalPlan(global_pose, global_plan);
+  const int goal_direction_ = base_local_planner::direction(*(global_plan.end() - std::min(3ul, global_plan.size())), global_plan);
```

Comment:
Q: is this even needed? Isn't it enough to just 
`return base_local_planner::direction(global_pose, global_plan) != -1;` ?

Reply:
`goal_direction_` can be either `-1` or `+1`, so your suggestion wouldn't work.

(btw just noticed it has an `_` in the name, will fix)

Reply:
So as far as I understand this, direction `-1` means past the goal, `0` means next to the goal, and `1` means before the goal. The `base_local_planner::direction` uses the 3rd last pose to check whether a point is past the goal. And the line here checks whether the 3rd last pose is past the goal -> this can never be the case? `goal_direction_` can only be 1 (or 0 if 3rd last pose somehow was equal to the last pose, e.g. if `global_plan.size()` is 1) ?

I.e. whether we passed the goal should be determinable by a single dot product of (goal - third last pose) with (goal - current pose): the goal was passed if this is <0  ?

Reply:
> goal_direction_ can only be 1 (or 0 if 3rd last pose somehow was equal to the last pose, e.g. if global_plan.size() is 1) ?

I'm not exactly sure what we're computing here, but this is not true (maybe @ayushgaud can explain better).

I tested by manually dragging the robot around while printing the value of `goal_direction`.
The value of `goal_direction` can be -1 and +1:

https://user-images.githubusercontent.com/15384781/187823217-21013995-1fbe-4af3-aa19-2e2bf2598f82.mp4

```cpp
bool LatchedStopRotateController::isGoalBypassed(LocalPlannerUtil* planner_util,
                                                 const geometry_msgs::PoseStamped& global_pose)
{
  std::vector<geometry_msgs::PoseStamped> global_plan;
  planner_util->getLocalPlan(global_pose, global_plan);
  const int goal_direction =
      base_local_planner::direction(*(global_plan.end() - std::min(3ul, global_plan.size())), global_plan);
  ROS_FATAL_STREAM(goal_direction);
  return base_local_planner::direction(global_pose, global_plan) != goal_direction;
}
```

Reply:
@ayushgaud is this expected? I.e. is the statement `direction -1 means past the goal, 0 means next to the goal, and 1 means before the goal. ` not correct?

In either case, I would just change the `direction` code to a dot product and if we want to keep the int returns `return dot_p == 0 ? 0 : std::copysign(1, dot_p)`. Then there's no need to store a  `goal_direction` variable, and in `isGoalBypassed`, we can just `return base_local_planner::direction(global_pose, global_plan) != -1;`

Reply:
Fixed